Rcpp::compileAttributes (dot once more)

This is related to #500 and a question on [stackoverflow from 2013](https://stackoverflow.com/questions/20720796/package-compilation-fails-when-package-name-contain-a-dot-and-rcpp-function). 

Assume you have a package with a dot in the name `with.dot`, and you want to provide a C++ interface. This causes an error upon compilation:


```RRcpp::Rcpp.package.skeleton('with.dot')
f <- file('./with.dot/src/rcpp_hello_world.cpp', 'a+')
cat('// [[Rcpp::interfaces(cpp)]]', file = f)
close(f)
wd <- setwd('./with.dot')
Rcpp::compileAttributes()
devtools::document(roclets = c('rd', 'collate', 'namespace'))
setwd(wd)
```
Here the compilation is executed by `devtools::document` and this causes an error:

```bash
Re-compiling with.dot
-  installing *source* package 'with.dot' ... (557ms)
   ** using staged installation
   ** libs
   "C:/RBuildTools/4.0/mingw64/bin/"g++ -std=gnu++11  -I"C:/Users/olive/R/R-cran/R-40~1.3/include" -DNDEBUG  -I'C:/Users/olive/Documents/R/win-library/4.0/Rcpp/include'        -O2 -Wall  -mfpmath=sse -msse2 -mstackrealign -UNDEBUG -Wall -pedantic -g -O0 -fdiagnostics-color=always -c RcppExports.cpp -o RcppExports.o
   In file included from RcppExports.cpp:4
   ../inst/include/with.dot.h:7:10:fatal error: dot_RcppExports.h: No such file or directory
    #include "with_dot_RcppExports.h"
             ^~~~~~~~~~~~~~~~~~~~~~~~
   compilation terminated.
   make: *** [C:/Users/olive/R/R-cran/R-40~1.3/etc/x64/Makeconf:230: RcppExports.o] Error 1
   ERROR: compilation failed for package 'with.dot'
-  removing 'C:/Users/olive/AppData/Local/Temp/RtmpUlqJPr/devtools_install_20b02c3c2615/with.dot'
Error in (function (command = NULL, args = character(), error_on_status = TRUE,  : 
  System command 'Rcmd.exe' failed, exit status: 1, stdout + stderr (last 10 lines):
E> ** libs
E> "C:/RBuildTools/4.0/mingw64/bin/"g++ -std=gnu++11  -I"C:/Users/olive/R/R-cran/R-40~1.3/include" -DNDEBUG  -I'C:/Users/olive/Documents/R/win-library/4.0/Rcpp/include'        -O2 -Wall  -mfpmath=sse -msse2 -mstackrealign -UNDEBUG -Wall -pedantic -g -O0 -fdiagnostics-color=always -c RcppExports.cpp -o RcppExports.o
E> In file included from RcppExports.cpp:4
E> ../inst/include/with.dot.h:7:10:fatal error: dot_RcppExports.h: No such file or directory
E>  #include "with_dot_RcppExports.h"
E>           ^~~~~~~~~~~~~~~~~~~~~~~~
E> compilation terminated.
E> make: *** [C:/Users/olive/R/R-cran/R-40~1.3/etc/x64/Makeconf:230: RcppExports.o] Error 1
E> ERROR: compilation failed for package 'with.dot'
E> * removing 'C:/Users/olive/AppData/Loc
[...]
Type .Last.error.trace to see where the error occured
In addition: Warning message:
roxygen2 requires Encoding: UTF-8 
```

The problem comes from the file `with.dot/inst/with.dot.h` trying to reference `with.dot/inst/with.dot_RcppExports.h` (note the dot in the file name). It seems to me, that file names should have replaced dots (`.`) by underscores (`_`) for reference purposes. I tried going through the debugger and got as far as 
```R
invisible(.Call("compileAttributes", PACKAGE = "Rcpp", 
        pkgdir, pkgname, depends, registration, cppFiles, cppFileBasenames, 
        includes, verbose, .Platform))
```
at which point my search stopped.


______
```R
sessionInfo()
R version 4.0.3 (2020-10-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=English_Denmark.1252  LC_CTYPE=English_Denmark.1252   
[3] LC_MONETARY=English_Denmark.1252 LC_NUMERIC=C                    
[5] LC_TIME=English_Denmark.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5        rstudioapi_0.13   xml2_1.3.2        knitr_1.30        roxygen2_7.1.1   
 [6] magrittr_2.0.1    usethis_2.0.0     devtools_2.3.2    pkgload_1.1.0     R6_2.5.0         
[11] rlang_0.4.9       fansi_0.4.1       stringr_1.4.0     tools_4.0.3       pkgbuild_1.2.0   
[16] xfun_0.19         sessioninfo_1.1.1 cli_2.2.0         withr_2.3.0       ellipsis_0.3.1   
[21] remotes_2.2.0     assertthat_0.2.1  digest_0.6.27     rprojroot_2.0.2   lifecycle_0.2.0  
[26] crayon_1.3.4      processx_3.4.5    purrr_0.3.4       callr_3.5.1       fs_1.5.0         
[31] ps_1.5.0          testthat_3.0.1    memoise_1.1.0     glue_1.4.2        stringi_1.5.3    
[36] compiler_4.0.3    desc_1.2.0        prettyunits_1.1.1
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,8 @@
+2021-01-17  Dirk Eddelbuettel  <edd@debian.org>
+
+        * DESCRIPTION (Version, Date): Roll minor version
+        * inst/include/Rcpp/config.h: Idem
+
 2021-01-16  Dirk Eddelbuettel  <edd@debian.org>
 
 	* tests/tinytest.R: Test for CODECOV when deciding to run extensive

--- a/ChangeLog
+++ b/ChangeLog
@@ -12,6 +12,8 @@
 	* R/RcppClass.R: Idem
 	* R/RcppLdpath.R: Idem
 
+	* src/attributes.cpp: Support include files with dots in package name
+
 2021-01-14  Dirk Eddelbuettel  <edd@debian.org>
 
         * DESCRIPTION (Date, Version): Release 1.0.6

--- a/DESCRIPTION
+++ b/DESCRIPTION
@@ -1,7 +1,7 @@
 Package: Rcpp
 Title: Seamless R and C++ Integration
-Version: 1.0.6
-Date: 2021-01-14
+Version: 1.0.6.1
+Date: 2021-01-17
 Author: Dirk Eddelbuettel, Romain Francois, JJ Allaire, Kevin Ushey, Qiang Kou,
  Nathan Russell, Douglas Bates and John Chambers
 Maintainer: Dirk Eddelbuettel <edd@debian.org>

--- a/inst/include/Rcpp/config.h
+++ b/inst/include/Rcpp/config.h
@@ -30,7 +30,7 @@
 #define RCPP_VERSION_STRING     "1.0.6"
 
 // the current source snapshot
-#define RCPP_DEV_VERSION        RcppDevVersion(1,0,6,0)
-#define RCPP_DEV_VERSION_STRING "1.0.6.0"
+#define RCPP_DEV_VERSION        RcppDevVersion(1,0,6,1)
+#define RCPP_DEV_VERSION_STRING "1.0.6.1"
 
 #endif

--- a/src/attributes.cpp
+++ b/src/attributes.cpp
@@ -633,6 +633,9 @@ namespace attributes {
         // if it wasn't (throws exception on io error)
         bool commit(const std::string& preamble = std::string());
 
+        // Convert a dot in package name to underscore for use in header file name
+        std::string dotNameHelper(const std::string & name) const;
+
     private:
 
         // Private virtual for doWriteFunctions so the base class
@@ -717,7 +720,7 @@ namespace attributes {
         std::string includeDir_;
     };
 
-    // Class which manages generating PackageName_RcppExports.h header file
+    // Class which manages generating PackageName.h header file
     class CppPackageIncludeGenerator : public ExportsGenerator {
     public:
         CppPackageIncludeGenerator(const std::string& packageDir,
@@ -1894,6 +1897,13 @@ namespace attributes {
         return removeFile(targetFile_);
     }
 
+    // Convert a possible dot in package name to underscore as needed for header file
+    std::string ExportsGenerator::dotNameHelper(const std::string & name) const {
+        std::string newname(name);
+        std::replace(newname.begin(), newname.end(), '.', '_');
+        return newname;
+    }
+
     CppExportsGenerator::CppExportsGenerator(const std::string& packageDir,
                                              const std::string& package,
                                              const std::string& fileSep)
@@ -2128,7 +2138,7 @@ namespace attributes {
                                             const std::string& fileSep)
         : ExportsGenerator(
             packageDir +  fileSep + "inst" +  fileSep + "include" +
-            fileSep + package + kRcppExportsSuffix,
+            fileSep + dotNameHelper(package) + kRcppExportsSuffix,
             package,
             "//")
     {
@@ -2346,7 +2356,7 @@ namespace attributes {
                                             const std::string& fileSep)
         : ExportsGenerator(
             packageDir +  fileSep + "inst" +  fileSep + "include" +
-            fileSep + package + ".h",
+            fileSep + dotNameHelper(package) + ".h",
             package,
             "//")
     {
@@ -2359,7 +2369,6 @@ namespace attributes {
             std::string guard = getHeaderGuard();			// #nocov start
             ostr() << "#ifndef " << guard << std::endl;
             ostr() << "#define " << guard << std::endl << std::endl;
-
             ostr() << "#include \"" << packageCpp() << kRcppExportsSuffix
                    << "\"" << std::endl;
 

